### PR TITLE
Update web-app.md

### PR DESCRIPTION
**Pull request by LongfeiH**
# Description

Add web-app content to web pages which moved from Microsoft tutorial with https://github.com/microsoft/ML-For-Beginners/tree/main/3-Web-App.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# How has this been tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

-  [local build the book](https://github.com/open-academy/machine-learning/blob/main/CONTRIBUTING.md#build-the-book)
-  review the artifacts from local build
-  [local build the slides](https://github.com/open-academy/machine-learning/blob/main/CONTRIBUTING.md#build-the-slides-optional) if appliable
- [ ] Unit test if appliable

# Checklist:

-  My code follows the [style guidelines](https://github.com/open-academy/machine-learning/blob/main/STYLE_GUIDE.md) of this project
-  I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
-  I have made corresponding changes to the documentation
-  My changes generate no new warnings
-  I have applied tests that prove my fix is effective or that my feature works
-  New and existing unit tests pass locally with my changes
-  Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by LongfeiH** [comment]
add assignments

File: open-machine-learning-jupyter-book/assignments/ml-fundamentals/web-app.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Try a different model
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/ml-fundamentals/web-app.md:1]: make the title to be consistent with the file name.

File: open-machine-learning-jupyter-book/assignments/ml-fundamentals/web-app.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+## Instructions
+
+Now that you have built one web app using a trained Regression model, use one of the models from an earlier Regression lesson to redo this web app. You can keep the style or design it differently to reflect the pumpkin data. Be careful to change the inputs to reflect your model's training method.
+
+## Rubric
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/assignments/ml-fundamentals/web-app.md:7]: You need to add the acknowledge section. E.g. https://open-academy.github.io/machine-learning/assignments/data-science/small-diabetes-study.html#acknowledgments

File: open-machine-learning-jupyter-book/ml-fundamentals/web-app.md [hunk 3]
```diff
@@ -1 +1,346 @@
-# Web app
\ No newline at end of file
+# Web app
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:1]: follow numpy.md to add jupytext header

File: open-machine-learning-jupyter-book/ml-fundamentals/web-app.md [hunk 4]
```diff
@@ -1 +1,346 @@
...
+# Build a Web App to use a ML Model
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:2]: I think let's use this as the title and you can update the file name to be consistent.

Also, make ML to be Machine Learning. And Model -> model

File: open-machine-learning-jupyter-book/ml-fundamentals/web-app.md [hunk 5]
```diff
@@ -1 +1,346 @@
...
+
+In this lesson, you will train an ML model on a data set that's out of this world: _UFO sightings over the past century_, sourced from NUFORC's database.
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:4]: We don't use lesson, instead using section to replace it.

File: open-machine-learning-jupyter-book/ml-fundamentals/web-app.md [hunk 6]
```diff
@@ -1 +1,346 @@
...
+You will learn:
+
+- How to 'pickle' a trained model
+- How to use that model in a Flask app
+
+We will continue our use of notebooks to clean data and train our model, but you can take the process one step further by exploring using a model 'in the wild', so to speak: in a web app.
+
+To do this, you need to build a web app using Flask.
+
+## [Pre-lecture quiz](https://gray-sand-07a10f403.1.azurestaticapps.net/quiz/17/)
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:15]: remove this.

File: open-machine-learning-jupyter-book/ml-fundamentals/web-app.md [hunk 7]
```diff
@@ -1 +1,346 @@
...
+## Exercise - clean your data
+
+In this lesson you'll use data from 80,000 UFO sightings, gathered by [NUFORC](https://nuforc.org) (The National UFO Reporting Center). This data has some interesting descriptions of UFO sightings, for example:
+
+- **Long example description.** "A man emerges from a beam of light that shines on a grassy field at night and he runs towards the Texas Instruments parking lot".
+- **Short example description.** "the lights chased us".
+
+The `ufos.csv` spreadsheet includes columns about the `city`, `state` and `country` where the sighting occurred, the object's `shape` and its `latitude` and `longitude`.
+
+In the blank `notebook` included in this lesson:
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:54]: where is the notebook?

Comment (zhangqi444) [reply open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:54]: As the code is executable here. Do we still need to create a notebook file?

File: open-machine-learning-jupyter-book/ml-fundamentals/web-app.md [hunk 8]
```diff
@@ -1 +1,346 @@
...
+✅ What's [Pickle](https://docs.python.org/3/library/pickle.html)? Pickle 🥒 is a Python module that serializes and de-serializes a Python object structure. When you 'pickle' a model, you serialize or flatten its structure for use on the web. Be careful: pickle is not intrinsically secure, so be careful if prompted to 'un-pickle' a file. A pickled file has the suffix `.pkl`.
+
+## Exercise - clean your data
+
+In this lesson you'll use data from 80,000 UFO sightings, gathered by [NUFORC](https://nuforc.org) (The National UFO Reporting Center). This data has some interesting descriptions of UFO sightings, for example:
+
+- **Long example description.** "A man emerges from a beam of light that shines on a grassy field at night and he runs towards the Texas Instruments parking lot".
+- **Short example description.** "the lights chased us".
+
+The `ufos.csv` spreadsheet includes columns about the `city`, `state` and `country` where the sighting occurred, the object's `shape` and its `latitude` and `longitude`.
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:52]: where is the `ufos.csv`? We need to have it in asserts/data

File: open-machine-learning-jupyter-book/ml-fundamentals/web-app.md [hunk 9]
```diff
@@ -1 +1,346 @@
...
+
+In the blank `notebook` included in this lesson:
+
+1. import `pandas`, `matplotlib`, and `numpy` as you did in previous lessons and import the ufos spreadsheet. You can take a look at a sample data set:
+
+    ```python
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:58]: use {code-cell} and please make sure the code below is executed.

File: open-machine-learning-jupyter-book/ml-fundamentals/web-app.md [hunk 10]
```diff
@@ -1 +1,346 @@
...
+    Your data should look like this:
+
+    ```output
+    	Seconds	Country	Latitude	Longitude
+    2	20.0	3		53.200000	-2.916667
+    3	20.0	4		28.978333	-96.645833
+    14	30.0	4		35.823889	-80.253611
+    23	60.0	4		45.582778	-122.352222
+    24	3.0		3		51.783333	-0.783333
+    ```
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:105]: We don't need such section if use {code-cell} for above

File: open-machine-learning-jupyter-book/ml-fundamentals/web-app.md (hunk 9 above)
Comment (zhangqi444) [reply open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:58]: Please apply to all the code if possible. But seems like some flask related code may not be executable. You can have a try.

File: open-machine-learning-jupyter-book/ml-fundamentals/web-app.md [hunk 11]
```diff
@@ -1 +1,346 @@
...
+1. The form variables are gathered and converted to a numpy array. They are then sent to the model and a prediction is returned.
+2. The Countries that we want displayed are re-rendered as readable text from their predicted country code, and that value is sent back to index.html to be rendered in the template.
+
+Using a model this way, with Flask and a pickled model, is relatively straightforward. The hardest thing is to understand what shape the data is that must be sent to the model to get a prediction. That all depends on how the model was trained. This one has three data points to be input in order to get a prediction.
+
+In a professional setting, you can see how good communication is necessary between the folks who train the model and those who consume it in a web or mobile app. In our case, it's only one person, you!
+
+---
+
+## 🚀 Challenge
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:334]: move this to Your turn section.

File: open-machine-learning-jupyter-book/ml-fundamentals/web-app.md [hunk 12]
```diff
@@ -1 +1,346 @@
...
+
+Instead of working in a notebook and importing the model to the Flask app, you could train the model right within the Flask app! Try converting your Python code in the notebook, perhaps after your data is cleaned, to train the model from within the app on a route called `train`. What are the pros and cons of pursuing this method?
+
+## [Post-lecture quiz](https://gray-sand-07a10f403.1.azurestaticapps.net/quiz/18/)
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:338]: remove

File: open-machine-learning-jupyter-book/ml-fundamentals/web-app.md [hunk 13]
```diff
@@ -1 +1,346 @@
...
+
+## Review & Self Study
+
+There are many ways to build a web app to consume ML models. Make a list of the ways you could use JavaScript or Python to build a web app to leverage machine learning. Consider architecture: should the model stay in the app or live in the cloud? If the latter, how would you access it? Draw out an architectural model for an applied ML web solution.
+
+## Assignment
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:344]: We call this as Your turn section. https://open-academy.github.io/machine-learning/data-science/working-with-data/relational-databases.html#your-turn

File: open-machine-learning-jupyter-book/ml-fundamentals/web-app.md [hunk 14]
```diff
@@ -1 +1,346 @@
...
+
+---
+
+## 🚀 Challenge
+
+Instead of working in a notebook and importing the model to the Flask app, you could train the model right within the Flask app! Try converting your Python code in the notebook, perhaps after your data is cleaned, to train the model from within the app on a route called `train`. What are the pros and cons of pursuing this method?
+
+## [Post-lecture quiz](https://gray-sand-07a10f403.1.azurestaticapps.net/quiz/18/)
+
+## Review & Self Study
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:340]: Please make the title consistent with other chapters, https://open-academy.github.io/machine-learning/data-science/working-with-data/relational-databases.html

Also, please keep the section in the same order as other chapters.

File: open-machine-learning-jupyter-book/ml-fundamentals/web-app.md [hunk 15]
```diff
@@ -1 +1,346 @@
...
+        return render_template(
+            "index.html", prediction_text="Likely country: {}".format(countries[output])
+        )
+    
+    
+    if __name__ == "__main__":
+        app.run(debug=True)
+    ```
+
+    > 💡 Tip: when you add [`debug=True`](https://www.askpython.com/python-modules/flask/flask-debug-mode) while running the web app using Flask, any changes you make to your application will be reflected immediately without the need to restart the server. Beware! Don't enable this mode in a production app.
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/ml-fundamentals/web-app.md:313]: use {note} https://jupyterbook.org/en/stable/content/content-blocks.html

**Review by zhangqi444 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**